### PR TITLE
treewide: Fix unsafe concatenation of $LD_LIBRARY_PATH, round 2

### LINKED ISSUE #144646: Correctly joining `LD_LIBRARY_PATH` and similar bash variables throughout nixpkgs
There are places in nixpkgs where the pattern `LD_LIBRARY_PATH=somedir:$LD_LIBRARY_PATH` is used.

That is a security risk (see  [article](https://jdhao.github.io/2021/07/03/ld_library_path_empty_item/)).

This issue is about removing all such bad patterns from nixpkgs, for `LD_LIBRARY_PATH`, `PATH`, and potentially other variables.

###### Security risk explanation

If `$LD_LIBRARY_PATH` hasn't been set before (is undefined or the empty string), then the variable expansion will end in `...:`, thus having an empty entry on the right side.

Having an empty entry in it puts the _current directory_ onto `LD_LIBRARY_PATH` (e.g.  `somedir:` is the same as `somedir:.`) , thus loading `.so` files and executing code that the user may not wish to execute.

For example, loading downloaded `.so` files from the Internet in a directory the user `cd`d into.

###### Nixpkgs example

Here is one such case that even wraps a packaged program permanently with such a wrong `LD_LIBRARY_PATH`:

https://github.com/NixOS/nixpkgs/blob/164173d0fc914ec5f70b67cf1c8b2f03142cb7fc/pkgs/applications/office/jabref/default.nix#L53-L55

## Correct variable concatenation in plain shell

There are 2 standard **correct patterns** that prepend the separator `:` only if `LD_LIBRARY_PATH`, or any variable `V`, is not itself empty:

1. Conditionally expanding `:$V` (the combination of `:` and `$V`), from https://unix.stackexchange.com/questions/162891/append-to-path-like-variable-without-creating-leading-colon-if-unset:
   ```sh
   export V="somethingPrepended${V:+:${V}}"
   export LD_LIBRARY_PATH="somedir${LD_LIBRARY_PATH:+:${LD_LIBRARY_PATH}}"
   ```
2. Conditionally expanding just the `:`, as currently more common in nixpkgs ([example for appending here](https://github.com/NixOS/nixpkgs/blob/164173d0fc914ec5f70b67cf1c8b2f03142cb7fc/pkgs/development/libraries/silgraphite/graphite2.nix#L39)):
   ```sh
   export V="somethingPrepended${V:+:}${V}"
   export LD_LIBRARY_PATH="somedir${LD_LIBRARY_PATH:+:}$LD_LIBRARY_PATH"
   ```

I personally think that both are quite unreadable, and easy to get wrong.

## Correct variable concatenation in nix bash scripts

Further, in nix multi-line string literals (`''`), escaping of `$` is needed, making it `''$`:

```nix
''
   export V="somethingPrepended''${V:+:''${V}}"
   export LD_LIBRARY_PATH="somedir''${LD_LIBRARY_PATH:+:''${LD_LIBRARY_PATH}}"

   export V="somethingPrepended''${V:+:}''${V}"
   export LD_LIBRARY_PATH="somedir''${LD_LIBRARY_PATH:+:}''$LD_LIBRARY_PATH"
''
```

The needed escaping makes it even more difficult to get it right, which probably contributes to this correct way not being used everywhere.

## What could be done to improve this?

One or multiple of the below:

* Provide a nix function that makes this easier (I suggest one below), and enforce it in reviews.
* Provide a shell function that makes this easier (similar to `substituteInPlace`).
* Enforce one of the above safe patterns in reviews (may be tough, as they are difficult to remember).
* Add a linter that can find such issues in reviews automatically.

### PR DESCRIPTION
**Pull request by andersk**
###### Motivation for this change

Naive concatenation of `$LD_LIBRARY_PATH` can result in an empty colon-delimited segment; this tells glibc to load libraries from the current directory, which is definitely wrong, and may be a security vulnerability if the current directory is untrusted.  (See #67234, for example.)  Fix this throughout the tree.

Followup to #76804.  Fixes #144646.

Cc @fricklerhandwerk @wamserma @danieldk @drewrisinger @danielbarter @adisbladis @kira-bruneau @nh2

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/openbabel-bindings/default.nix [hunk 1]
```diff
@@ -13,7 +13,7 @@ buildPythonPackage rec {
   # some functionality to work (inparticular, pybel).
   # see https://openbabel.org/docs/dev/Installation/install.html
   BABEL_LIBDIR = "${openbabel}/lib/openbabel/3.1.0";
-  LD_LIBRARY_PATH = "$LD_LIBRARY_PATH:${openbabel}/lib";
+  LD_LIBRARY_PATH = "${openbabel}/lib";
```
Comment (andersk) [inline pkgs/development/python-modules/openbabel-bindings/default.nix:16]: The previous version just put the literal unexpanded string `$LD_LIBRARY_PATH` into the `LD_LIBRARY_PATH` variable, which wasn’t useful.

```console
$ nix-shell -A python3Packages.openbabel-bindings --run 'echo "$LD_LIBRARY_PATH"'
$LD_LIBRARY_PATH:/nix/store/hc33pncm3ggml00kay5h64fkivxnsdni-openbabel-3.1.1/lib
```

File: pkgs/tools/games/gamemode/preload-nix-workaround.patch [hunk 2]
```diff
@@ -6,7 +6,7 @@ index 573b3e4..6f2799e 100755
  
  # ld will find the right path to load the library, including for 32-bit apps.
  LD_PRELOAD="${GAMEMODEAUTO_NAME}${LD_PRELOAD:+:$LD_PRELOAD}"
-+LD_LIBRARY_PATH="@libraryPath@${LD_LIBRARY_PATH:+:$LD_LIBRARY_PATH}"
++LD_LIBRARY_PATH="@libraryPath@${LD_LIBRARY_PATH:+:}$LD_LIBRARY_PATH"
```
Comment (andersk) [inline pkgs/tools/games/gamemode/preload-nix-workaround.patch:9]: (This was fine either way, but the latter avoids false positives when grepping for `:\$LD_LIBRARY_PATH` to find real problems.)

**Comment by r-burns** [comment]
Changes LGTM

How do you feel about using addToSearchPath where possible? Long term, I think encouraging use of a safe helper function will be easier than enforcing this relatively esoteric bash syntax.

**Comment by andersk** [comment]
`addToSearchPath` has a few differences that made me less confident in using it as a drop-in replacement:

* It’s only available at build time and not at runtime.
* It only adds to the end of the path.
* It checks that the argument is an existing directory, so you can’t use it to, e.g., add many directories at once.
* It’s an undocumented helper function in `setup.sh`, so I’m not sure if it was intended to be a public interface.

If someone wants to work on adding more uses of `addToSearchPath` or something like it, I’m not opposed, but that should probably be another PR.

**Comment by r-burns** [comment]
Okay, sounds good. Let's defer that discussion for now.

**Review by r-burns — APPROVED** [review]

**Pull request merged**